src/CommandLine.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef COMMANDLINE_H
+#define COMMANDLINE_H
+
+#include "sbuf/SBuf.h"
+#include <vector>
+
+/// Manages arguments passed to the program, including the program name.
+/// The same info is passed to main() as argc and argv[] parameters.
+class CommandLine
+{
+public:
+    // codes for options without short option characters
+    enum LongCodes {
+        ForegroundCode = 1,
+        KidCode = 2
+    };
+
+    typedef std::pair<char, SBuf> OptionsPair;
+    typedef std::list<OptionsPair> Options;
+
+    CommandLine(int argc, char *argv[]);
+
+    /// \returns parsed kid option argument or an empty string
+    SBuf kidName() const;
+
+    /// apply all available command line options
+    void processOptions();
+
+    /// generate a new argument list from the parsed one,
+    /// supstituting argv[0] and adding/substituting kid option
+    const char **argv(const char *argv0, const char *kidName);
+
+    /// \returns Squid executable file name
```

Comment:
That's a lie. `argv[0]` is often not a Squid executable file name. If there is no "standard" name for `argv[0]`, let's call this argZero(), arg0(), or programName().

Reply:
I disagree. This method is used for providing the first argument to the execvp() system call. The man page documentation says:
```
The initial argument for these functions is the name of a file that is to be executed.
```
So that is namely the 'name of executable file', not something else. The fact that execFile_ is initialized from argv[0] does not matter: we do not have another way to initialize it.

Reply:
> This method is used for providing the first argument to the execvp() system call.

First of all, when documenting a method, we should focus on what the method does, not on how it is (mis)used. Doing so is difficult and sometimes nearly impossible, but not in this case AFAICT.

Second, while the first argument to execvp() system call has to be an executable name, `argv[0]` is not necessarily an executable file name (e.g., it is something else for Squid kids).

Most likely, CommandLine should deal with `argv` in general and `argv[0]` semantics in particular rather than dabble into execvp() parameters. It is possible that CommandLine should deal with both (i.e., store both `argv[0]` and the executable name), but I doubt it is a good idea. The correct answer here may become clear once the other/bigger/obvious scoping problems are fixed. Let's come back to this then if needed.